File: sample/python/ShakeBodies.py
from cnoid.Util import *
from cnoid.Base import *
from cnoid.BodyPlugin import *
from numpy import *

class ShakeBodies:
    def __init__(self):
        toolBar = ToolBar("ShakeBar")
        toolBar.setVisibleByDefault(True)
        self.button = toolBar.addToggleButton("Shake")
        self.button.setChecked(True)
        self.button.toggled.connect(self.onButtonToggled)
        MainWindow.getInstance().addToolBar(toolBar)

        self.bodyItems = []
        self.dp = array([0.0, 0.0, 0.01])
        self.connections = ScopedConnectionSet()
        self.connections.add(
            ItemTreeView.getInstance().getSigSelectionChanged().connect(self.onSelectionChanged))
        self.timer = Timer()
        self.connections.add(
            self.timer.timeout.connect(self.onTimeout))

    def onButtonToggled(self, on):
        self.onSelectionChanged(ItemTreeView.getInstance().getSelectedItems())

    def onSelectionChanged(self, items):
        self.bodyItems = BodyItemList(items)
        if len(self.bodyItems) > 0 and self.button.isChecked():
            self.timer.start(50)
        else:
            self.timer.stop()

    def onTimeout(self):
        for bodyItem in self.bodyItems:
            body = bodyItem.getBody()
            body.getRootLink().p += self.dp
            body.calcForwardKinematics()
            bodyItem.notifyKinematicStateChange()
        self.dp = -self.dp

shakeBodies = ShakeBodies()

File: sample/lua/ShakeBodies.lua
cnoid.require "Util"
cnoid.require "Base"
cnoid.require "Body"
cnoid.require "BodyPlugin"

BodyShaker = {
   new = function()
      local self = { }
      self.toolBar = cnoid.ToolBar.new("ShakeBar")
      self.toolBar:setVisibleByDefault(true)
      self.button = self.toolBar:addToggleButton("Shake")
      self.button:setChecked(true)

      self.bodyItems = { }
      self.dp = cnoid.Vector3(0.0, 0.0, 0.01)
      self.connections = cnoid.ScopedConnectionSet.new()
      self.timer = cnoid.Timer.new()

      function onSelectionChanged(items)
	 self.bodyItems = cnoid.extractBodyItems(items)
	 if #self.bodyItems > 0 and self.button:isChecked() then
	    self.timer:start(50)
	 else
	    self.timer:stop()
	 end
      end

      self.connections:add(
	 cnoid.ItemTreeView.instance():sigSelectionChanged():connect(onSelectionChanged))

      self.connections:add(
	 self.button:sigToggled():connect(
	    function(on)
	       onSelectionChanged(cnoid.ItemTreeView.instance():selectedItems())
	    end))

      self.onTimeout = function()
	 for i, bodyItem in ipairs(self.bodyItems) do
	    local body = bodyItem:body()
	    body:rootLink():setTranslation(body:rootLink():translation() + self.dp)
	    body:calcForwardKinematics()
	    bodyItem:notifyKinematicStateChange()
	    self.dp = -self.dp
	 end
      end

      self.connections:add(
	 self.timer:sigTimeout():connect(self.onTimeout))

      cnoid.MainWindow.instance():addToolBar(self.toolBar)

      setmetatable(self, { 
         __gc = function(self)
	    print "cnoid.MainWindow.instance():removeToolBar(self.toolBar)"
	    cnoid.MainWindow.instance():removeToolBar(self.toolBar)
	 end
      })

      return self
   end
}

bodyShaker = BodyShaker.new()
